src/Git/Value/SemVerVersion.php

```diff
@@ -24,16 +23,14 @@ private function __construct(int $major, int $minor, int $patch)
 
     /**
      * @psalm-pure
-     * @psalm-suppress ImpureFunctionCall the {@see \Safe\preg_match()} API is pure by design
+     * @psalm-suppress ImpureFunctionCall {@see https://github.com/azjezz/psl/issues/130}
      */
     public static function fromMilestoneName(string $name): self
     {
-        Assert::notEmpty($name);
-        Assert::regex($name, '/^(v)?\\d+\\.\\d+\\.\\d+$/');
-
-        preg_match('/(\\d+)\\.(\\d+)\\.(\\d+)/', $name, $matches);
+        Psl\invariant(Regex\matches($name, '/^(v)?\\d+\\.\\d+\\.\\d+$/'), 'Milestone name is malformed.');
 
-        Assert::isList($matches);
+        /** @var array{1: string, 2: string, 3: string} $matches */
```

Comment:
Can coercion improve here perhaps? :thinking:

Reply:
psalm already can detect `array{1: string, 2: string, 3: string}|null` return type, i will move the assertion above below this call and check for null instead.